drivers/video/video_stm32_dcmi.c

```diff
@@ -0,0 +1,511 @@
+	.enqueue = video_stm32_dcmi_enqueue,
+	.dequeue = video_stm32_dcmi_dequeue,
+	.get_caps = video_stm32_dcmi_get_caps,
+};
+
+static void video_stm32_dcmi_irq_config_func(const struct device *dev)
+{
+	IRQ_CONNECT(DT_INST_IRQN(0), DT_INST_IRQ(0, priority),
+		stm32_dcmi_isr, DEVICE_DT_INST_GET(0), 0);
+	irq_enable(DT_INST_IRQN(0));
+}
+
+#define DCMI_DMA_CHANNEL_INIT(index, src_dev, dest_dev)					\
+	.dma_dev = DEVICE_DT_GET(DT_INST_DMAS_CTLR_BY_IDX(index, 0)),			\
+	.channel = DT_INST_DMAS_CELL_BY_IDX(index, 0, channel),				\
+	.reg = (DMA_TypeDef *)DT_REG_ADDR(						\
+				DT_PHANDLE_BY_IDX(DT_DRV_INST(0), dmas, 0)),		\
+	.cfg = {									\
+		.dma_slot = STM32_DMA_SLOT_BY_IDX(index, 0, slot),			\
+		.channel_direction = STM32_DMA_CONFIG_DIRECTION(			\
+			STM32_DMA_CHANNEL_CONFIG_BY_IDX(index, 0)),			\
+		.source_data_size = STM32_DMA_CONFIG_##src_dev##_DATA_SIZE(		\
+			STM32_DMA_CHANNEL_CONFIG_BY_IDX(index, 0)),			\
+		.dest_data_size = STM32_DMA_CONFIG_##dest_dev##_DATA_SIZE(		\
+			STM32_DMA_CHANNEL_CONFIG_BY_IDX(index, 0)),			\
+		.source_burst_length = 1,       /* SINGLE transfer */			\
+		.dest_burst_length = 1,         /* SINGLE transfer */			\
+		.channel_priority = STM32_DMA_CONFIG_PRIORITY(				\
+			STM32_DMA_CHANNEL_CONFIG_BY_IDX(index, 0)),			\
+		.dma_callback = dmci_dma_callback,					\
+	},										\
+
+PINCTRL_DT_INST_DEFINE(0);
+
+#define STM32_DCMI_GET_CAPTURE_RATE(capture_rate)					\
+	((capture_rate) == 1 ? DCMI_CR_ALL_FRAME :					\
+	(capture_rate) == 2 ? DCMI_CR_ALTERNATE_2_FRAME :				\
+	(capture_rate) == 4 ? DCMI_CR_ALTERNATE_4_FRAME :				\
+	DCMI_CR_ALL_FRAME)
+
+#define STM32_DCMI_GET_BUS_WIDTH(bus_width)						\
+	((bus_width) == 8 ? DCMI_EXTEND_DATA_8B :					\
+	(bus_width) == 10 ? DCMI_EXTEND_DATA_10B :					\
+	(bus_width) == 12 ? DCMI_EXTEND_DATA_12B :					\
+	(bus_width) == 14 ? DCMI_EXTEND_DATA_14B :					\
+	DCMI_EXTEND_DATA_8B)
+
+#define DCMI_DMA_CHANNEL(id, src, dest)							\
+	.dma = {									\
+		COND_CODE_1(DT_INST_DMAS_HAS_IDX(id, 0),				\
+			(DCMI_DMA_CHANNEL_INIT(id, src, dest)),				\
+			(NULL))								\
+	},
+
+static struct video_stm32_dcmi_data video_stm32_dcmi_data = {
+	.hdcmi = {
+		.Instance = (DCMI_TypeDef *) DT_INST_REG_ADDR(0),
+		.Init = {
+				.SynchroMode = DCMI_SYNCHRO_HARDWARE,
+				.PCKPolarity = (DT_INST_PROP(0, pixelclk_active) ?
+						DCMI_PCKPOLARITY_RISING : DCMI_PCKPOLARITY_FALLING),
+				.HSPolarity = (DT_INST_PROP(0, hsync_active) ?
+						DCMI_HSPOLARITY_HIGH : DCMI_HSPOLARITY_LOW),
+				.VSPolarity = (DT_INST_PROP(0, vsync_active) ?
+						DCMI_VSPOLARITY_HIGH : DCMI_VSPOLARITY_LOW),
+				.CaptureRate = STM32_DCMI_GET_CAPTURE_RATE(
+							DT_INST_PROP(0, capture_rate)),
+				.ExtendedDataMode = STM32_DCMI_GET_BUS_WIDTH(
+							DT_INST_PROP(0, bus_width)),
+				.JPEGMode = DCMI_JPEG_DISABLE,
+				.ByteSelectMode = DCMI_BSM_ALL,
+				.ByteSelectStart = DCMI_OEBS_ODD,
+				.LineSelectMode = DCMI_LSM_ALL,
+				.LineSelectStart = DCMI_OELS_ODD,
+		},
```

Comment:
Do these needs to be in RAM? the code does not change them, maybe they can be moved to config structure (though I suspect the HAL ma need them in ram)

Another thing to consider is to initialize these structure in runtime, may save a bit of flash by not having to store an entire copy of of `video_stm32_dcmi_data` in flash since you don't need to statically initialize all of it, probably not a big deal for the SoC class this stuff runs of but something to consider

Reply:
Moved `struct stream dma` to config as suggested. 

Regarding `video_stm32_dcmi_data`, as it only initializes `hdcmi`, I prefer to keep it as is, based on other STM32 drivers. Is that okay for you?

Reply:
Yeah just wanted to make sure you considered the option, personally I try to leave the data structures zero initialized as much as possible but as I said since this is a soc specific driver and these are usually on the high end of the specs I think it's fine. If it was a general purpose driver it'd be more important.